source/common/router/router.cc

```diff
@@ -348,19 +382,28 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
           if (!config_.suppress_envoy_headers_) {
             headers.insertEnvoyOverloaded().value(Http::Headers::get().EnvoyOverloadedValues.True);
           }
+          if (modify_headers_ != nullptr) {
+            modify_headers_(headers);
+          }
         },
         absl::nullopt, StreamInfo::ResponseCodeDetails::get().MaintenanceMode);
     cluster_->stats().upstream_rq_maintenance_mode_.inc();
     return Http::FilterHeadersStatus::StopIteration;
   }
 
   // Fetch a connection pool for the upstream cluster.
-  Http::ConnectionPool::Instance* conn_pool = getConnPool();
-  if (!conn_pool) {
+  conn_pool_ = getConnPool();
```

Comment:
I think it turns out I don't need this change. I was saving the pool as a member variable so I could reference it later in `onUpstreamHeaders()`, but now it's just captured by `modify_headers_`. Will revert this part.